axum-extra/src/either.rs

```diff
@@ -0,0 +1,233 @@
+
+use axum::{
+    async_trait,
+    extract::{FromRequest, RequestParts},
+    response::{IntoResponse, Response},
+};
+
+/// Combines two extractors or responses into a single type.
+///
+/// See the [module docs](self) for examples.
+#[derive(Debug, Clone)]
+pub enum Either2<E1, E2> {
+    #[allow(missing_docs)]
+    E1(E1),
+    #[allow(missing_docs)]
+    E2(E2),
+}
+
+/// Combines three extractors or responses into a single type.
+///
+/// See the [module docs](self) for examples.
+#[derive(Debug, Clone)]
+pub enum Either3<E1, E2, E3> {
+    #[allow(missing_docs)]
+    E1(E1),
+    #[allow(missing_docs)]
+    E2(E2),
+    #[allow(missing_docs)]
+    E3(E3),
+}
+
+/// Combines four extractors or responses into a single type.
+///
+/// See the [module docs](self) for examples.
+#[derive(Debug, Clone)]
+pub enum Either4<E1, E2, E3, E4> {
+    #[allow(missing_docs)]
+    E1(E1),
+    #[allow(missing_docs)]
+    E2(E2),
+    #[allow(missing_docs)]
+    E3(E3),
+    #[allow(missing_docs)]
+    E4(E4),
+}
+
+/// Combines five extractors or responses into a single type.
+///
+/// See the [module docs](self) for examples.
+#[derive(Debug, Clone)]
+pub enum Either5<E1, E2, E3, E4, E5> {
+    #[allow(missing_docs)]
+    E1(E1),
+    #[allow(missing_docs)]
+    E2(E2),
+    #[allow(missing_docs)]
+    E3(E3),
+    #[allow(missing_docs)]
+    E4(E4),
+    #[allow(missing_docs)]
+    E5(E5),
+}
+
+/// Combines six extractors or responses into a single type.
+///
+/// See the [module docs](self) for examples.
+#[derive(Debug, Clone)]
+pub enum Either6<E1, E2, E3, E4, E5, E6> {
+    #[allow(missing_docs)]
+    E1(E1),
+    #[allow(missing_docs)]
+    E2(E2),
+    #[allow(missing_docs)]
+    E3(E3),
+    #[allow(missing_docs)]
+    E4(E4),
+    #[allow(missing_docs)]
+    E5(E5),
+    #[allow(missing_docs)]
+    E6(E6),
+}
+
+/// Combines seven extractors or responses into a single type.
+///
+/// See the [module docs](self) for examples.
+#[derive(Debug, Clone)]
+pub enum Either7<E1, E2, E3, E4, E5, E6, E7> {
+    #[allow(missing_docs)]
+    E1(E1),
+    #[allow(missing_docs)]
+    E2(E2),
+    #[allow(missing_docs)]
+    E3(E3),
+    #[allow(missing_docs)]
+    E4(E4),
+    #[allow(missing_docs)]
+    E5(E5),
+    #[allow(missing_docs)]
+    E6(E6),
+    #[allow(missing_docs)]
+    E7(E7),
+}
+
+/// Combines eight extractors or responses into a single type.
+///
+/// See the [module docs](self) for examples.
+#[derive(Debug, Clone)]
+pub enum Either8<E1, E2, E3, E4, E5, E6, E7, E8> {
+    #[allow(missing_docs)]
+    E1(E1),
+    #[allow(missing_docs)]
+    E2(E2),
+    #[allow(missing_docs)]
+    E3(E3),
+    #[allow(missing_docs)]
+    E4(E4),
+    #[allow(missing_docs)]
+    E5(E5),
+    #[allow(missing_docs)]
+    E6(E6),
+    #[allow(missing_docs)]
+    E7(E7),
+    #[allow(missing_docs)]
+    E8(E8),
+}
+
+macro_rules! impl_traits_for_either {
+    (
+        $either:ident =>
+        [$($ident:ident),* $(,)?],
+        $last:ident $(,)?
```

Comment:
I think this can probably be simplified, but I can play around with that after this PR is merged.